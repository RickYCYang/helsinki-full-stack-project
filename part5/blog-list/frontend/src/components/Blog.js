import React, { useState } from 'react';
import PropTypes from 'prop-types';

const Blog = ({ blog, updateBlog, removeBlog, user }) => {
  const blogStyle = {
    paddingTop: 10,
    paddingLeft: 2,
    border: 'solid',
    borderWidth: 1,
    marginBottom: 5,
  };
  const [showDetail, setShowDetail] = useState(false);
  const isPoster = blog.user.id === user.id;

  return (
    <li style={blogStyle} className="blog">
      <span>
        {blog.title} {blog.author}
      </span>
      <button onClick={() => setShowDetail(!showDetail)}>
        {showDetail ? 'hide' : 'view'}
      </button>
      {showDetail && (
        <div>
          <p>{blog.url}</p>
          <p>
            likes {blog.likes}{' '}
            <button onClick={() => updateBlog(blog)}>like</button>
          </p>
          <p>{blog.user.username}</p>
          {isPoster && (
            <p>
              <button onClick={() => removeBlog(blog)}>remove</button>
            </p>
          )}
        </div>
      )}
    </li>
  );
};

Blog.propTypes = {
  blog: PropTypes.object.isRequired,
  updateBlog: PropTypes.func.isRequired,
  removeBlog: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
};

export default Blog;
