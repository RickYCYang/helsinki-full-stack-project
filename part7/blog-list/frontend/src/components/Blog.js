import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useMatch } from 'react-router-dom';
import { fetchBlogs, updateBlog } from '../redux/reducers/blogsReducer';
import BlogComments from './BlogComment';
import { isEmpty } from 'lodash';
import Button from '@mui/material/Button';

const Blog = () => {
  const dispatch = useDispatch();
  const match = useMatch('/blogs/:id');
  const blog = useSelector((state) =>
    state.blogs.find((blog) => blog.id === match.params.id)
  );

  useEffect(() => {
    if (isEmpty(blog)) {
      dispatch(fetchBlogs());
    }
  }, []);

  if (isEmpty(blog)) return null;

  return (
    <div>
      <h2>{blog.title}</h2>
      <a href={blog.url}>{blog.url}</a>
      <p>
        {blog.likes} likes{' '}
        <Button variant="contained" onClick={() => dispatch(updateBlog(blog))}>
          like
        </Button>
      </p>
      <p>added by {blog.user.username}</p>
      <BlogComments blog={blog} />
    </div>
  );
};

export default Blog;
