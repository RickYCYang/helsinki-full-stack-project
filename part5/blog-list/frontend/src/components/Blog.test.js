import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Blog from './Blog';

/** userEvent v14 issue reported link:
 * https://github.com/testing-library/user-event/issues/839
 *
 */

const blogId = '62dba9f463c7037a6f57c051';
const title = 'Component testing is done with react-testing-library';
const author = 'Aris';
const url = 'www.google.com';
const likes = 10;
const userAuth = {
  id: 1,
  name: 'Rick',
  username: 'Rick',
  token: 'token',
};

const blog = {
  id: blogId,
  title: title,
  url: url,
  author: author,
  likes: likes,
  user: {
    id: userAuth.id,
    name: userAuth.name,
    username: userAuth.username,
  },
};

describe('<Blog />', () => {
  const mockUpdateBlog = jest.fn();
  const mockRemoveBlog = jest.fn();
  let container;

  beforeEach(() => {
    container = render(
      <Blog
        blog={blog}
        user={userAuth}
        updateBlog={mockUpdateBlog}
        removeBlog={mockRemoveBlog}
      />
    ).container;
  });

  afterEach(() => {
    cleanup();
  });

  test('renders content', () => {
    const element = screen.getByText(`${blog.title} ${blog.author}`);
    expect(element).toBeDefined();
    expect(screen.queryByText(blog.likes)).toBeNull();
    expect(screen.queryByText(blog.url)).toBeNull();

    const div = container.querySelector('.blog');
    expect(div).toHaveTextContent(
      'Component testing is done with react-testing-library Aris'
    );
  });

  test('check url and number of likes are shown after clicking show button', async () => {
    const viewButton = screen.getByText('view');
    await userEvent.click(viewButton);

    const urlLabel = screen.getByText(blog.url);
    expect(urlLabel).toBeInTheDocument();
    const likesLabel = screen.getByText(`likes ${blog.likes}`);
    expect(likesLabel).toBeInTheDocument();
  });

  test('clicking the button calls event handler once', async () => {
    const viewButton = screen.getByText('view');
    let likeButton = screen.queryByText('like');
    let removeButton = screen.queryByText('remove');
    expect(likeButton).not.toBeInTheDocument();
    expect(removeButton).not.toBeInTheDocument();

    await userEvent.click(viewButton);
    //screen.debug();

    likeButton = screen.queryByText('like');
    removeButton = screen.queryByText('remove');
    const hideButton = screen.getByText('hide');
    expect(likeButton).toBeInTheDocument();
    expect(removeButton).toBeInTheDocument();
    expect(hideButton).toBeInTheDocument();

    await userEvent.click(likeButton);
    await userEvent.click(removeButton);
    expect(mockUpdateBlog).toHaveBeenCalledTimes(1);
    expect(mockRemoveBlog).toHaveBeenCalledTimes(1);
    await userEvent.click(likeButton);
    expect(mockUpdateBlog).toHaveBeenCalledTimes(2);
  });

  test('snapshot of html elements', () => {
    expect(container).toMatchSnapshot();
  });
});
