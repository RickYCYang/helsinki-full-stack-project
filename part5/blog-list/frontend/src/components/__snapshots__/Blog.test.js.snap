// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Blog /> snapshot of html elements 1`] = `
<div>
  <li
    class="blog"
    style="padding-top: 10px; padding-left: 2px; border: solid; border-width: 1px; margin-bottom: 5px;"
  >
    <span>
      Component testing is done with react-testing-library
       
      Aris
    </span>
    <button>
      view
    </button>
  </li>
</div>
`;
